test/paymentsSpec.js

```diff
@@ -114,6 +115,124 @@ describe('Payments', function () {
     sinon.assert.calledWith(fulfillSpy, '130394ed-f621-4663-80dc-910adc66f4c6', 'HS8e5Ew02XKAglyus2dh2Ohabuqmy3HDM8EXMLz22ok')
   })
 
+  it('should retry condition fulfillment', function * () {
+    const stub = sinon.stub(this.mockPlugin2, 'fulfillCondition')
+
+    // faking the clock triggers the retrying logic faster and makes the test run faster
+    let inTestCase = true
+    const interval = this.setInterval(() => {
+      if (inTestCase) {
+        this.clock.tick(1000)
+      }
+    }, 5)
+
+    stub.onFirstCall().returns(Promise.reject('first fulfillment attempt failed'))
+    stub.onSecondCall().returns(Promise.reject('second fulfillment attempt failed'))
+
+    yield this.mockPlugin1.emitAsync('outgoing_fulfill', {
+      id: '800c25e5-5db4-4cde-8d34-72edad1601a8',
+      direction: 'outgoing',
+      noteToSelf: {
+        source_transfer_id: '130394ed-f621-4663-80dc-910adc66f4c6',
+        source_transfer_ledger: 'mock.test2.',
+        source_transfer_amount: '1.0'
+      },
+      expiresAt: (new Date(START_DATE + 5000)).toISOString()
+    }, 'HS8e5Ew02XKAglyus2dh2Ohabuqmy3HDM8EXMLz22ok')
+
+    sinon.assert.calledThrice(stub)
+    sinon.assert.calledWith(stub, '130394ed-f621-4663-80dc-910adc66f4c6', 'HS8e5Ew02XKAglyus2dh2Ohabuqmy3HDM8EXMLz22ok')
+
+    // clean-up the fake timer to avoid side-effects with other tests
+    clearInterval(interval)
+    inTestCase = false
+  })
+
+  it('should not retry on certain errors', function * () {
+    // these errors should not trigger a retry
+    const errorNames = ['InvalidFieldsError', 'TransferNotFound', 'AlreadyRolledBackError',
+      'TransferNotConditionalError', 'NotAcceptedError']
+
+    // faking the clock triggers the retrying logic faster and makes the test run faster
+    let inTestCase = true
+    const interval = this.setInterval(() => {
+      if (inTestCase) {
+        this.clock.tick(1000)
+      }
+    }, 5)
+
+    for (const name of errorNames) {
+      const stub = sinon.stub(this.mockPlugin2, 'fulfillCondition')
+      const e = new Error()
+      e.name = name
+      stub.onFirstCall().returns(Promise.reject(e))
+
+      yield this.mockPlugin1.emitAsync('outgoing_fulfill', {
+        id: '800c25e5-5db4-4cde-8d34-72edad1601a8',
+        direction: 'outgoing',
+        noteToSelf: {
+          source_transfer_id: '130394ed-f621-4663-80dc-910adc66f4c6',
+          source_transfer_ledger: 'mock.test2.',
+          source_transfer_amount: '1.0'
+        },
+        expiresAt: (new Date(START_DATE + 5000)).toISOString()
+      }, 'HS8e5Ew02XKAglyus2dh2Ohabuqmy3HDM8EXMLz22ok')
+
+      sinon.assert.calledOnce(stub)
+      sinon.assert.calledWith(stub, '130394ed-f621-4663-80dc-910adc66f4c6', 'HS8e5Ew02XKAglyus2dh2Ohabuqmy3HDM8EXMLz22ok')
+      stub.restore()
+    }
+    // clean-up the fake timer to avoid side-effects with other tests
+    clearInterval(interval)
+    inTestCase = false
+  })
+
+  it('should timeout retrying condition fulfillment', function * () {
```

Comment:
This test only passes if the retry logic times out as expected. If it would not time out, the connector would keep retrying the fulfillment until the mocha test times out, which would let the test fail. You can test that yourself by commenting out the timeout promise in `executeSourceTransfer`.

I added an additional assertion that checks that the fulfillment was attempted the correct number of times.